### DISCUSSION TITLE
Allow method on types definition inside structs

### DISCUSSION THREAD

**Discussion by mp81ss** [vlang/v#11695] [Ideas]
It would be very nice to allow something like this:

```
struct Abc {
    name string
    // all other fields

    fn (a Abc) is_empty() bool {
        return a.name.len == 0
    }
    // all other funcs
}
```

This imply the use of type Abc before its complete definition.
I would add method definition inside structs AFTER all fields have been declared, as shown above.
This would greatly simulate OOP without changing the language

**Comment by baxiry** [comment]
In this case, ```(a Abc)``` will not be required. like this:
```
struct Abc {
    name string

    fn  is_empty() bool {
        return this.name.len == 0
    }
}
```
When I was learning Go and Rust, I was wondering about the benefit of separating the method from the body struct.
Their merging into one body struct seemed more friendly and clear in reading. I do not know.

But I got used to this. I don't think it is so important

**Comment by mp81ss** [comment]
Yes, merging methods inside parent data-structure is more friendly and clear in reading.
That's why I asked for it.
Your synthax with this keyword is equivalent and shorter, but requires changing the actual parser.
Both ways would be good

**Comment by StunxFS** [comment]
I like the idea, I'm used to that way of writing methods (from C#, C++, etc ...), it would be great if you could do that in V.
BTW, in addition to `this` you could use `self` like in Rust/Python.

```v
struct Abc {
    name string

    // like in Python (without `self` parameter)
    fn is_empty() bool {
        return self.name.len == 0
    }

    // or like in Rust
    fn is_empty(self) bool {
        return self.name.len == 0
    }

    fn is_empty_mut(mut self) bool {
        return self.name.len == 0
    }

    fn is_empty_ref(&self) bool {
        return self.name.len == 0
    }
}
```

**Comment by JalonSolov** [comment]
Why do you need to declare the functions inside the struct?  This already works:
```v
struct Abc {
    name string
    // all other fields
}

fn (a Abc) is_empty() bool {
    return a.name.len == 0
}
// all other funcs
```

Since it references `a`, it will be a per-instance reference.  No need for `self` or other placeholders.

**Comment by ntrel** [comment]
> In this case, (a Abc) will not be required. 

But sometimes we need a receiver of type `&Abc` or a mutable receiver.

**Comment by baxiry** [comment]
@ntrel You are right

**Comment by soldate** [comment]
struct Abc {
    name string

    fn is_empty() bool {
        return this.name.len == 0
    }
}

is much better than 

struct Abc {
    name string
}

fn (a Abc) is_empty() bool {
    return a.name.len == 0
}

i think the first way should be the mandatory or at least possible to do. 
The main principle of this language is simplicity.. do the cleaner and more obvious way to express what you want.

**Reply by JalonSolov** [reply]
No, it's not "better".  It's different.

structs are not classes, and do not have methods defined within.

They have receiver methods declared outside.  Different, not better or worse.

Just because you're used to something, doesn't make it the "best" way to do things.

**Comment by soldate** [comment]
sorry but for me 
fn is_empty() bool // with this or self
is better, cleaner, more obvious than
fn (a Abc) is_empty() bool

maybe it could be outside, but enforce to create a v file only for a struct.

== Abc.v ==

struct Abc {
name string
}

fn is_empty() bool {
return this.name.len == 0
}

plus this way is familiar for OOP (java?) programmers. Get them to became v programmers. :-)

**Reply by JalonSolov** [reply]
There is nothing stopping you from putting the struct and all receiver methods in their own file.

Receiver methods are perfectly familiar to Go programmers, as Go also only supports receiver methods.

**Reply by soldate** [reply]
There are much more java, python or javascript programmers than go programmers. get them all to v.

**Reply by Wajinn** [reply]
Different languages have their own way of doing objects.  JavaScript is a prototype-based OOP language, not class-based like Java.  JavaScript, to the disagreement of many users, allowed classes into the language, which is the source of continual controversy for them.

The Vlang way has a line of thinking and tradition, which comes from Oberon-2 (Niklaus Wirth) -> Golang -> Vlang.  This style is just as valid and has its advantages.  Other languages also do things differently, like Lua, which is prototype based like JavaScript.  The thought process shouldn't be that there is only a set way of doing things, which must be applied to all other languages.

**Comment by soldate** [comment]
look this code below (examples/http_server.v)

the struct ExampleHandler  was create empty. (???)
For what? To implement a method "handle" that starts with "fn (h ExampleHandler) "
to... not use "h" at all!

for me (java programmer), should be a better way to do this. this.

-----------------------------------------------------------------------------------
```v
module main

import net.http { CommonHeader, Request, Response, Server }

struct ExampleHandler {}

fn (h ExampleHandler) handle(req Request) Response {
	mut res := Response{
		header: http.new_header_from_map({
			CommonHeader.content_type: 'text/plain'
		})
	}
	mut status_code := 200
	res.body = match req.url {
		'/foo' {
			'bar\n'
		}
		'/hello' {
			'world\n'
		}
		'/' {
			'foo\nhello\n'
		}
		else {
			status_code = 404
			'Not found\n'
		}
	}
	res.status_code = status_code
	return res
}

fn main() {
	mut server := Server{
		handler: ExampleHandler{}
	}
	server.listen_and_serve()
}
```
